tests/conftest.py

```diff
@@ -74,40 +79,51 @@ async def sqlite_engine() -> AsyncEngine:
     return engine
 
 
-@pytest.fixture(params=["sqlite_session", "postgres_session"])
-def session(request) -> AsyncSession:
-    return request.getfixturevalue(request.param)
+@pytest.fixture
+def session(request, dialect) -> AsyncSession:
+    if dialect == "sqlite":
+        return request.getfixturevalue("sqlite_session")
+    elif dialect == "postgresql":
+        return request.getfixturevalue("postgres_session")
+    raise ValueError(f"Unknown session fixture: {dialect}")
 
 
-@pytest.fixture(params=["sqlite_db", "postgres_db"])
-def db(request) -> async_sessionmaker:
-    return request.getfixturevalue(request.param)
+@pytest.fixture
+def db(request, dialect) -> async_sessionmaker:
+    if dialect == "sqlite":
+        return request.getfixturevalue("sqlite_db")
+    elif dialect == "postgresql":
+        return request.getfixturevalue("postgres_db")
+    raise ValueError(f"Unknown db fixture: {dialect}")
 
 
 @pytest.fixture
-async def sqlite_db(sqlite_engine: AsyncEngine) -> Callable[[], AsyncContextManager[AsyncSession]]:
+async def sqlite_db(
+    sqlite_engine: AsyncEngine,
+) -> AsyncGenerator[Callable[[], AsyncContextManager[AsyncSession]], None]:
     Session = async_sessionmaker(sqlite_engine, expire_on_commit=False)
 
-    @contextlib.asynccontextmanager
-    async def factory() -> AsyncIterator[AsyncSession]:
-        async with Session.begin() as session:
+    async with Session.begin() as session:
+
+        @contextlib.asynccontextmanager
+        async def factory() -> AsyncIterator[AsyncSession]:
             yield session
 
-    return factory
+        yield factory
```

Comment:
This change just means that the transaction can no longer end _inside_ the test function, since it can only end when the test function exits, correct? It probably doesn't make a difference, but just want to confirm that's the intent here.

Reply:
yeah, it is, it's so that all of our database-related fixtures share the same revertable transaction. This way we can pass in the engine and fixtures to some tests, and the session and fixtures to lower level tests, and they are all referencing the same database state

Reply:
If that's the case, then it's somewhat misleading, since using it like below seems to suggest that the transaction is ended when the context manager exits, but in fact it does not.

```python
async with db() as session:
    ...
```

Reply:
> they are all referencing the same database state

Why is that beneficial? Not that it matters, unless the tests are executing data manipulation, but if that is true, would that make the tests interdependent?

Reply:
On the other hand, since the fixture is function-scoped, this doesn't change anything.

Reply:
This is beneficial because it leaves open the possibility of simultaneously maintaining test isolation (fixtures are fresh per test and only loaded if asked for) and someday supporting module or session scoped database tests. If we structure this so that transactions actually commit to the database I think we'll have to revisit all the tests that use that pattern if we ever move away from function-scoped databases

Reply:
> because it leaves open the possibility

Not sure if I fully understand this rationale, since we will have do [dynamic scope](https://docs.pytest.org/en/8.2.x/how-to/fixtures.html#dynamic-scope) anyway, so this is all moot.